[ v2.6-branch] net: sockets: tls: Support for DER cert chain and NOCOPY optimisation

lucasdietrich opened the pull request:
net: sockets: tls: Support for DER cert chain and NOCOPY optimisation

Add TLS socket option "TLS_CERT_NOCOPY" to prevent the copy of 
certificates to mbedTLS heap if possible.
Add an option in MQTT client to take advantage of the new 
"TLS_CERT_NOCOPY" option.

Add support to provide a chain of DER certificates by registering 
them with multiple tags.

rlubos commented on subsys/net/lib/mqtt/mqtt_transport_socket_tls.c:
```diff
@@ -78,6 +78,15 @@ int mqtt_client_tls_connect(struct mqtt_client *client)
 		}
 	}
 
+	if (tls_config->cert_nocopy) {
```
How about
```suggestion
	if (tls_config->cert_nocopy != TLS_CERT_NOCOPY_NONE ) {
```

rlubos commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -756,25 +805,24 @@ static int tls_set_credential(struct tls_context *tls,
 			      struct tls_credential *cred)
 {
 	switch (cred->type) {
+
```
I don't think this empty line is needed.

rlubos commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -756,25 +805,24 @@ static int tls_set_credential(struct tls_context *tls,
+
+	case TLS_CREDENTIAL_PRIVATE_KEY:
 	{
 		struct tls_credential *priv_key =
```
Not sure I follow this one, can't we just set `tls_set_private_key(tls, cred)`? The currently processed credential is the private key, I don't see a reason to look it up again.

rlubos commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -836,8 +889,21 @@ static int tls_mbedtls_set_credentials(struct tls_context *tls)
+			 */
+			if (!priv_key_found) {
+				err = -ENOENT;
+				goto exit;
```
I don't think this is right, going back to exit will cause `credentials_unlock()` being called twice. Anyway, I kind of lean towards what you wrote in the related issue - I guess we could just rely on mbed TLS here to produce an error in case own certificate is registered w/o a private key (so here we'd only check for `own_cert_present`).

rlubos reviewed (commented):
Looks pretty neat. One more suggestion though, I'd consider moving MQTT related changes to a separate commit, as they're not tightly related.

Adecy commented:
Thanks @rlubos, I think I fixed the points you requested.

This PR is based on branch 2.6 because our application is using Zephyr-2.6, but I checked if this fix could be apply to the latest main branch. The only conflict is related to function mbedtls_pk_parse_key() which changed of signature.
I will not be able to properly test the changes myself, but should I rebase this PR over latest branch anyway ?

rlubos commented:
> Thanks @rlubos, I think I fixed the points you requested.
> 
> This PR is based on branch 2.6 because our application is using Zephyr-2.6, but I checked if this fix could be apply to the latest main branch. The only conflict is related to function mbedtls_pk_parse_key() which changed of signature. I will not be able to properly test the changes myself, but should I rebase this PR over latest branch anyway ?

Yes we should aim to get this into the main branch. I'm actually not even sure if we allow non-bugfix changes into release branches, @carlescufi @nashif what's the policy here?

lucasdietrich committed:
net: sockets: tls: Support for DER cert chain and NOCOPY optimisation

Add TLS socket option "TLS_CERT_NOCOPY" to prevent the copy of
certificates to mbedTLS heap if possible.

Add support to provide a chain of DER certificates by registering
them with multiple tags.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

lucasdietrich committed:
net: mqtt: Add support for TLS option TLS_CERT_NOCOPY

Add an option in MQTT client context to take advantage of the
"TLS_CERT_NOCOPY" option when using  TLS socket transport.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

carlescufi commented:
> Thanks @rlubos, I think I fixed the points you requested.
> 
> This PR is based on branch 2.6 because our application is using Zephyr-2.6, but I checked if this fix could be apply to the latest main branch. The only conflict is related to function mbedtls_pk_parse_key() which changed of signature. I will not be able to properly test the changes myself, but should I rebase this PR over latest branch anyway ?

Thanks for the PR @Adecy. Please send a PR to `main` and then also keep this one open in parallel. We need the feature first in main and then it can be merged to the 2.6-branch.
